### PR TITLE
Kris/reassign host

### PR DESCRIPTION
**Pull request by kristabel-wong**
- reassigned host
- modal for confirming exit or re-assigning host
- update of host number sits in firestore
- creation of playerNumber has been updated - more robust (takes the highest player number from player array, irregardless of order and uses that for player number creation

### CODE REVIEW

File: src/Components/Lobby.tsx [hunk 1]
```diff
@@ -17,11 +17,11 @@ const Lobby = () => {
   const playerDispatch = usePlayerDispatch();
 
   const [playerName, setPlayerName] = useState("");
-  const [isHost, setIsHost] = useState(false);
   const [promptCode, setPromptCode] = useState(false);
   const [existingRoomCode, setExistingRoomCode] = useState("");
   const [failJoinRoomMessage, setFailJoinRoomMessage] = useState("");
   const [alert, setAlert] = useState<boolean>(false)
+  const [currentPlayer, setCurrentPlayer] = useState({})
```
Comment (victorzw895) [inline src/Components/Lobby.tsx:23]: I noticed youre using currentPlayer just for the number value right?
we should be able to skip using a localState for this.
if you take a look at `usePlayerDocState` hook
it returns an object with these values. (dont worry about useMemo, its just to improve performance)
```
{players, setPlayers, player, setPlayer}
```

the player object should also have the number value
and it currently gets set once gameStarts (`src/Components/Board.tsx` line 40)

^ is still necessary as it updates player directions and abilities.
But you can update the player state earlier when creating/joining room.

File: src/Components/Lobby.tsx [hunk 2]
```diff
@@ -37,17 +37,19 @@ const Lobby = () => {
...
   const createNewGame = async () => {
     const newGameCode = cryptoRandomString({length: 5, type: 'distinguishable'});
-    setIsHost(true);
+    // setIsHost(true);
     // save Room Code
     gameDispatch({type: "joinRoom", value: newGameCode});
     // create new player
     const {player, dbPlayer}: PlayerFactoryType = PlayerFactory(playerName, 0)
 
+    setCurrentPlayer(player)
```
Comment (victorzw895) [inline src/Components/Lobby.tsx]: here you can do `setPlayer(dbPlayer)` instead

`setPlayer` coming from `usePlayerDocState`

File: src/Components/Lobby.tsx [hunk 3]
```diff
@@ -89,6 +93,8 @@ const Lobby = () => {
           players: playersInRoom
         },
       )
+      console.log("setting current player in join room", player)
+      setCurrentPlayer(player)
```
Comment (victorzw895) [inline src/Components/Lobby.tsx]: here you can do setPlayer(dbPlayer) instead

File: src/Components/Lobby.tsx [hunk 4]
```diff
@@ -108,7 +114,7 @@ const Lobby = () => {
       </h3>
       
       {gameState.roomId ?
-        <WaitingRoom isHost={isHost} />
+        <WaitingRoom currentPlayer={currentPlayer}/>
```
Comment (victorzw895) [inline src/Components/Lobby.tsx]: with `usePlayerDocState` this wont be necessary

File: src/Components/WaitingRoom.tsx [hunk 5]
```diff
@@ -1,19 +1,31 @@
...
-import {deleteDoc} from "firebase/firestore";
+import { deleteDoc, updateDoc } from "firebase/firestore";
+import CloseIcon from '@mui/icons-material/Close';
+import { useRoomHostDocState } from '../Contexts/FirestoreContext';
 
-const WaitingRoom = ({isHost}: {isHost: boolean}) => {
+const WaitingRoom = ({currentPlayer}: {currentPlayer: any}) => {
   const { gameState, gameDispatch } = useGame();
 
+  const [open, setOpen] = useState<boolean>(false)
```
Comment (victorzw895) [inline src/Components/WaitingRoom.tsx:16]: you can add here `const { player } = usePlayerDocState()`
currentPlayer would be player with extra values

File: src/Components/WaitingRoom.tsx [hunk 6]
```diff
@@ -38,39 +50,123 @@ const WaitingRoom = ({isHost}: {isHost: boolean}) => {
...
+    p: 4,
+  };
+
+  const exitRoom = async () => {
+    if (players.length === 1) {
+      deleteRoom()
+    } else {
+      // ! Feel like there could be async issues below but need revision
+      // * remove current player from player list 
+      const currentPlayerIndex: number = players.findIndex((player: any) => player.number === currentPlayer.number)
```
Comment (victorzw895) [inline src/Components/WaitingRoom.tsx]: There is a nicer way you can do this. :)
line 71-74
const updatePlayers = players.filter(dbPlayer => dbPlayer.number !== player.number)

File: src/Components/WaitingRoom.tsx [hunk 7]
```diff
@@ -38,39 +50,123 @@ const WaitingRoom = ({isHost}: {isHost: boolean}) => {
...
+      console.log("current player index",currentPlayerIndex)
+      const updatePlayers = [...players]
+      updatePlayers.splice(currentPlayerIndex, 1)
+      console.log("updated players", updatePlayers)
+      
+      // changing host number
+      if (currentPlayer.number === host) {
+        await updateDoc(doc(gameState.roomId), {
```
Comment (victorzw895) [inline src/Components/WaitingRoom.tsx]: this is fine, but you can also achieve the same thing by
using the helper methods in `src/utils/useFirestore.tsx`

```
await setDoc(gameState.roomId, {
      host: updatePlayers[0].number,
      players: updatePlayers
  })
```

less imports

File: src/Components/Lobby.tsx (hunk 2 above)
Comment (victorzw895) [reply src/Components/Lobby.tsx]: as for the player value that comes from PlayerFactory, im looking to refactor this out eventually, as i am still using it in some places, but seems redundant

File: src/Components/Lobby.tsx [hunk 8]
```diff
@@ -74,7 +75,10 @@ const Lobby = () => {
 
     // if found
     if (roomFound && !roomFound.gameStarted && roomFound.players.length <= 8) {
-      const {player, dbPlayer}: PlayerFactoryType = PlayerFactory(playerName, roomFound.players.length);
+      // add player Number
+      let newPlayerNo = Math.max(...roomFound.players.map( obj => obj.number));
```
Comment (victorzw895) [inline src/Components/Lobby.tsx:83]: FYI Kris, i think this wont quite solve it. as you can run into this issue below
Where player number expands beyond 8

I reckon we may as well refactor the playerFactory code.
Ive got a solution, that ill run you through later

Screen Shot 2023-03-01 at 10 57 11 am
Screen Shot 2023-03-01 at 10 57 21 am

File: src/Components/WaitingRoom.tsx (hunk 6 above)
Comment (kristabel-wong) [reply src/Components/WaitingRoom.tsx]: chef's kiss

File: src/Components/WaitingRoom.tsx (hunk 7 above)
Comment (kristabel-wong) [reply src/Components/WaitingRoom.tsx:79]: hmmm i tried this but it didn't work for me :'(. Did you try it too?

File: src/Components/WaitingRoom.tsx (hunk 6 above)
Comment (kristabel-wong) [reply src/Components/WaitingRoom.tsx]: oh actually i think this didn't work for me. `filter()` only provides a shallow copy, but the spread operator creates a deep copy which then destructively splices the player out of this version

File: src/Components/WaitingRoom.tsx (hunk 7 above)
Comment (kristabel-wong) [reply src/Components/WaitingRoom.tsx]: 🫡 yes bosssss

File: src/Components/Lobby.tsx (hunk 8 above)
Comment (kristabel-wong) [reply src/Components/Lobby.tsx:83]: hmmm i see... i guess now it depends if we want to reshuffle all playerNumbers when a player leaves or allow for playerNumber to be any number. 

at least the array of players does not exceed 8 players

File: src/Components/WaitingRoom.tsx (hunk 6 above)
Comment (kristabel-wong) [reply src/Components/WaitingRoom.tsx]: ahhh yep i made it a deep copy
`const updatePlayers = [...players].filter(dbPlayer => dbPlayer.number !== currentPlayer.number)`

Comment (victorzw895) [reply src/Components/WaitingRoom.tsx]: yea i see what you mean.
yes, filter wont destructively change your original array. but you can save its value and use that, which will give you a new array of values.
so

```
const players = [{number: 1, number: 2 , number: 3}]
players.filter(player => player !== 2)

console.log(players) // --> [{number: 1, number: 2 , number: 3}]
```
 vs (notice how im saving the filtered values into updatedPlayers)
```
const players = [{number: 1, number: 2 , number: 3}]
const updatedPlayers = players.filter(player => player !== 2)

console.log(players) // --> [{number: 1, number: 2 , number: 3}]
console.log(updatedPlayers) // --> [{number: 1, number: 3}]
```

plus
```
const players = [{number: 1, number: 2 , number: 3}]
const updatedPlayers = [...players]
// saving a copy of players in updatedPlayers and then doing destructive operations on newPlayers would be the same memory wise as
// Just saving a shallow copy of a filtered array into updatedPlayers
```

Comment (kristabel-wong) [reply src/Components/WaitingRoom.tsx]: ahhh yep you are right. I think i forgot to change your original line:
`const updatePlayers = players.filter(dbPlayer => dbPlayer.number !== player.number)`
from `player.number` to `currentPlayer.number`

**Review by victorzw895 — APPROVED** [review]
Looking good :D

**Pull request merged**